Cce analytic test exe

moxcodes opened the pull request:
## Proposed changes

Adds the CCE analytic test executable, as well as utilities for checking the outputs of the h5 file and a couple of fixes of errors in the analytic solutions that were found when verifying the outputs of the executable.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Dependencies

- [x] #2608

moxcodes committed:
Make CCE analytic solutions specify InitializeJ

- The functional form for the most fitting initial data depends on which
  analytic solution being tested, so the analytic solutions must specify
  the initial data generator that works best for the solution in
  question.

moxcodes committed:
Fix errors in TeukolskyWave analytic solution

The AnalyticTestCharacteristicExtract executable added a couple of
commits later more thoroughly verifies that this analytic solution
is now correct.

moxcodes committed:
Add observation for expected News

moxcodes committed:
Add analytic test executable

- Also the necessary input files to perform basic versions of the tests:
  - BouncingBlackHole
  - GaugeWave
  - RotatingSchwarzschild
  - TeukolskyWave

nilsdeppe commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,207 @@
+    # we reserve the first and second command-line arguments for our own use.
+    # Those will be the input file and the run directory, respectively.
+    if (len(sys.argv) < 3):
+        print("Usage:\n" + "python CheckH5.py InputFile.yaml run/directory/ " +
```
could we use `argparse` for parsing the arguments? then we could do something like `python --input-file BLAH --rundir run/dir --pytest-args`?

nilsvu replied:
Yes I think we can do that with [parse_known_args](https://docs.python.org/3.9/library/argparse.html#partial-parsing). This seems to work:

```py
import unittest

class TestCase(unittest.TestCase):
    def test_greeting(self):
        self.assertEqual(self.greeting, "Hello World")

if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('--greeting')
    duplicate_test_case, remaining_args = parser.parse_known_args(
        namespace=TestCase)
    del duplicate_test_case
    unittest.main(argv=[parser.prog] + remaining_args)
```

This way you can access the CLI args in your test case just with `self.`

nilsdeppe commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,207 @@
+    # True if `test data` is approximately equal to `expected_data`, otherwise
+    # False. It is expected that both inputs to the function are types for
+    # which `abs` returns a float
+    def approx(self, test_data, expected_data):
```
does `unit_test` not have a way to do this comparison that we could use instead?

moxcodes replied:
There is an `assertAlmostEqual`, but as far as I could tell, it only supports a crude version of absolute tolerance.

nilsdeppe replied:
Yea... This seems to be another case of libraries not implementing relative tolerances (correctly/at all)... Maybe add a brief comment to the dox stating that so nobody accidentally changes it in the future?

nilsvu replied:
Can you use [numpy.testing.assert_allclose](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose)? (it supports relative tolerances)

In other tests we `import numpy.testing as npt`, so I suggest to do that here too.

nilsdeppe commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,207 @@
+        if isinstance(h5_file[test_entity], h5py.Dataset):
+            # It is recommended to include subTest commands once python3 can be
+            # required:
+            # with self.unit_test.subTest(test_entity=test_entity,
```
@nilsleiffischer is there a way to "hack" this into python2?

nilsvu replied:
Yea, we can install and import `unittest2` https://pypi.org/project/unittest2/. Either we `try` to `import unittest2` before falling back to `unittest` or we also use `unittest2` on Py3. I suppose the former would be preferable if the APIs of the two packages are sufficiently close.

nilsdeppe commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,207 @@
+            # required:
+            # with self.unit_test.subTest(test_entity=test_entity,
+            #                             expected_entity=expected_entity):
+            test_data = h5_file[test_entity][()]
```
for my own knowledge, I do not understand how to parse `[()]`... Help?

moxcodes replied:
That 'unwraps' a dataset into a numpy array.

nilsdeppe replied:
huh, that's a lot cleaner than what I've been doing (`np.asarray`), though arguably harder to understand. Either way, cool!

nilsdeppe commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,207 @@
+                        or test_data.dtype == np.dtype(np.complexfloating)):
+                    for index in np.ndindex(test_data.shape):
+                        # It is recommended to include subTest commands once
+                        # python3 can be required:
```
what we would do with subtests?  Could you elaborate a bit more on what exactly this will do?

moxcodes replied:
The subtests would ensure that:
- any failing test would be annotated with the dataset and index that failed
- the tests would continue after a failing item in the dataset was found

nilsdeppe replied:
:+1: I like that :)

nilsdeppe commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,207 @@
+                        # python3 can be required:
+                        # with self.unit_test.subTest(index=index):
+                        self.unit_test.assertTrue(
+                            self.approx(test_data[index], 0.0))
```
since we have an absolute and relative tolerance why not compare using that? Maybe I don't understand the use-case here

moxcodes replied:
Sorry, I'm confused by this comment -- the approx method uses the absolute and relative tolerance, this is just here to avoid code duplication.
A test that gives only a single value to compare against zero doesn't give a good measure of scale, so currently will use only the absolute tolerance value and the user is expected to set that to something reasonable for their problem. I can change that behavior, though, if you think something else would be better suited.

nilsdeppe replied:
yea, just ignore it, I didn't read well :(

nilsdeppe commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,207 @@
+    # `unit_test` must be the unit test object -- this class is to be
+    # constructed within a test case, and must have access to the
+    # calling unit test object so that it can access the assert and
+    # subTest member functions.
```
could you document all the arguments? I don't know what `test_h5_entity` should be and `expected_h5_entity`. It appears these differ with what exactly is being tested (a group, dataset or single number)

nilsdeppe commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,207 @@
+    # Perform the unit test comparisons between the dataset or group
+    # represented by the string `test_entity` and `expected_entity`.
+    # If `expected_entity` is `None`, then `test_entity` (or, if it is a group,
+    # the datasets within the group) is compared to 0.0.
```
with or without an absolute tolerance?

nilsdeppe commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,207 @@
+import h5py
+
+
+# Describes a particular comparison between H5 datasets or groups
```
why not make all these comments docstrings? I'd like to (eventually) pull all these sort of files into Doxygen

nilsvu replied:
@moxcodes When writing the docstrings, would you mind formatting them according to Google's Python style guide https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings? That's not in our guidelines (yet), but I'd recommend it :)

nilsdeppe commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,207 @@
+    # If neither the expected nor test objects are present in an h5 file, that
+    # file is skipped. However, if none of the h5 files have the requested
+    # datasets, the test fails.
+    # It is not supported for the `run_directory` to have h5 files from
```
sorry what do you mean here? Is this going to cause problems for tests run in parallel? I'm worried that this script will open a file and start looking through it, then a parallel test deletes the file under it because the other test "owns" the file. This should also be documented

nilsdeppe commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,207 @@
+    # multiple runs.
+    def perform_check(self, run_directory):
+        checks_performed = False
+        for filename in os.listdir(run_directory):
```
could we use `glob` here? I don't know if that's in the python standard library (would like to not depend on more modules if possible)

moxcodes replied:
glob would work, I think -- I'm neutral as to which library function to use.

nilsdeppe replied:
let's get @nilsleiffischer's opinion :)

nilsvu replied:
I like `glob` here since it's how we would specify the filenames on the command line as well. It's in the standard library.

I also agree with @nilsdeppe (if I understood the other review comments correctly) that it would be good to specify the glob in the input file, instead of reading all files ending in `.h5`.

nilsdeppe commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,207 @@
+                # It is recommended to include subTest commands once
+                # python3 can be required:
+                # with self.unit_test.subTest(filename=filename):
+                with h5py.File(filename, 'r') as check_h5:
```
This is going to give many failures on machines with many cores where a lot of tests are run in parallel, see comment above

moxcodes replied:
I've altered the cmake to ensure that each input file test that uses this script does so in a directory labelled by the input file name -- is that not sufficient? I suppose if we tried to run multiple copies of the same test than it would still blow up...

nilsdeppe replied:
ah okay, sorry I missed that. Mostly, though something like `PlaneWave1D` (ScalarWave) could occur in other systems, like curved scalar wave. There are others, especially in the hydro systems where I could see input file name overlap. For example, a smooth flow input file could conceivably exist for Newtonian hydro, rel hydro, and GRMHD. Conceivably similar solutions exist for other hydro (or hydro extension) systems. Would those cause problems?

nilsdeppe commented on tools/CheckH5.py:
````diff
@@ -0,0 +1,207 @@
+    # ```
+    # # H5Checks:
+    # #   - CheckLabel: "label"
+    # #     TestH5: "/h5_name.dat"
````
We should be forced to specify a glob to match the file name. That would fix the current bug when running in parallel.

nilsdeppe commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,207 @@
+import unittest
+import numpy as np
+import sys
+import yaml
```
this adds the yaml module as a dependency (or maybe it was previously through something else?), but in either case that needs to be added to the installation instructions

nilsdeppe commented on cmake/AddInputFileTests.cmake:
```diff
@@ -47,6 +47,15 @@ function(add_single_input_file_test INPUT_FILE EXECUTABLE CHECK_TYPE TIMEOUT)
       # Make sure we run the test in the build directory for cleaning its output
       WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
       )
+  elseif("${CHECK_TYPE}" STREQUAL "execute_check_h5")
```
we should add to the `Having your Contributions Merged into SpECTRE` section a `Testing Input Files` section. Would you be willing to either write this documentation or file an issue so we remember to add it (please mark the issue with the `documentation` label)?

moxcodes replied:
Sure, I can write that documentation.

moxcodes replied:
Opened #2780  -- I'll do it in another PR to avoid further cluttering this one.

nilsdeppe commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,207 @@
+        h5_check_list = []
+        parsed_yaml = yaml.safe_load(h5_checks_lines)
+        for check_block in parsed_yaml['H5Checks']:
+            print(check_block)
```
left over from debugging? Otherwise maybe some more info?

nilsdeppe reviewed (changes requested):
I only looked at the one commit so far since I think it's most useful to get feedback on that quick (?). Overall I think this is a reasonable design :)

nilsvu commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,207 @@
+            while lines[line_index][:6] == "#   - " or lines[
+                    line_index][:6] == "#     ":
+                h5_checks_lines += lines[line_index][2:]
+                line_index += 1
```
How about this:

```suggestion
def read_h5_checks_config_lines(input_filename):
    with open(input_filename, 'r') as input_file:
        lines = input_file.readlines()
        found_config = False
        for i, line in enumerate(lines):
            if line.strip() == '# H5Checks:':
                found_config = True
                yield line[2:]
                continue
            if found_config:
                if line.startswith('#   '):
                    yield line[2:]
                else:
                    return
        raise RuntimeError("Could not find...")

h5_checks_config_lines = read_h5_checks_config_lines(
    self.input_filename)
h5_checks_config = yaml.safe_load(''.join(h5_checks_config_lines))
```

nilsvu commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,207 @@
+            # It is recommended to include subTest commands once python3 can be
+            # required:
+            # with self.subTest(test=h5_check.test_h5_label):
+            h5_check.perform_check(run_directory)
```
Unless I'm missing something I don't think you need the `H5Check` class. It seems you're using it like a dictionary. How about this:

```py
def check_h5_files(unit_test, label, abs_tol):
    # Do your checks

for config in h5_checks_config['H5Checks']:
    check_h5_files(
        self,
        label=config['Label'],
        abs_tol=config['AbsoluteTolerance'])
```

You can also stick the `check_h5_files` function in the test case class, then you don't need to pass `self` along.

moxcodes replied:
Sure, I feel I frequently encounter cases in python where you can just drop a nested builtin data structure in, instead of writing a class. In a certain sense, any class with member variables can be made a dictionary...
However, I have found programming in that way frustrating in the end -- I prefer classes with the implied encapsulation (even if python doesn't seem to care much for encapsulation).
If you insist, I'll change it

nilsvu commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,207 @@
+            # required:
+            # with self.unit_test.subTest(test_entity=test_entity,
+            #                             expected_entity=expected_entity):
+            test_data = h5_file[test_entity][()]
```
Do you need to convert to numpy at all? The dataset behaves like a numpy array in many ways, and once you slice it you have an array anyway. I suggest you try to just remove the `[()]`. If you need it after all, I suggest you use `.value` instead of the empty slice.

moxcodes replied:
Can you be more explicit with the `.value` suggestion?  `h5_file[test_entity].value()` doesn't appear to work. And using the dataset alone slows down the test by a factor of 2-3, I assume because the `[()]` pulls the whole thing from disk at once.

moxcodes commented:
I've rebased this on develop, it is no longer dependent, and I've posted fixups for the reviews of the h5 checking utility.
Cheers!

nilsdeppe commented on cmake/AddInputFileTests.cmake:
```diff
@@ -170,4 +179,21 @@ ${Python_EXECUTABLE} ${CMAKE_SOURCE_DIR}/tools/CleanOutput.py -v \
+  WRITE
+  ${PROJECT_BINARY_DIR}/tmp/InputFileExecuteCheckH5AndClean.sh
+  "\
+#!/bin/sh\n\
```
should this have `-e` added? I don't remember what that's needed to correctly stop the intermediate steps in a script (or if that's a bash-ism)

nilsdeppe reviewed (commented):
LGTM, please go ahead and squash once @nilsleiffischer has taken a look :)

nilsdeppe commented:
@nilsleiffischer can you please look at the fixups?

nilsvu commented on src/Evolution/Systems/Cce/Actions/InitializeFirstHypersurface.hpp:
```diff
@@ -14,8 +14,13 @@
 #include "Parallel/GlobalCache.hpp"
 #include "Utilities/Gsl.hpp"
 #include "Utilities/TMPL.hpp"
+#include "Utilities/TypeTraits/IsA.hpp"
```
Could you clarify the commit message? I don't understand what you mean with "The best initial data can depend on the represented analytic solution"

nilsvu replied:
ping on this

nilsvu commented on cmake/AddInputFileTests.cmake:
```diff
@@ -170,4 +183,21 @@ ${Python_EXECUTABLE} ${CMAKE_SOURCE_DIR}/tools/CleanOutput.py -v \
+# Write command to execute an input file and clean its output into a shell
+# script, which makes it easier to chain multiple commands
+file(
+  WRITE
```
I think it would be preferable to make this a `configure_file` and put the script (including placeholder cmake variables) in `tools/` or `cmake/`. The reason is that `configure_file` doesn't re-write the file on every run of `cmake`, but only when it changes (if I understand the cmake docs correctly). Whether this is relevant for cmake runtime I don't know.

nilsvu commented on docs/Installation/Installation.md:
```diff
@@ -43,6 +43,7 @@ environment differently, read on!
 * [NumPy](http://www.numpy.org/) 1.10 or later
 * [SciPy](https://www.scipy.org)
 * [matplotlib](https://matplotlib.org/)
+* [PyYAML](https://pyyaml.org/)
```
This is currently installed in the container only as a dependency of `pybtex`. Please add it to the Dockerfile.

nilsvu commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,228 @@
+if sys.version_info < (3, 0):
+    import unittest2 as unittest
+else:
+    import unittest
```
Just do:

```py
# For Py2 compatibility
try:
    import unittest2 as unittest
except:
    import unittest
```

nilsvu commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,228 @@
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--input_filename')
+    parser.add_argument('--run_directory')
```
Please replace `_` -> `-` in these CLI arg names. The Python variables will still have the `_`.

nilsvu commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,228 @@
+            h5_check_list += [
+                H5Check(self, check_block.get("CheckLabel"),
+                        check_block.get("H5Glob"), check_block.get("TestH5"),
+                        check_block.get("ExpectedH5"),
```
(optional) A few suggestions for names:
- `CheckH5.py` -> `CheckOutputFiles.py`
- `H5Checks` -> `OutputFileChecks`
- `CheckLabel` -> `Label`
- `H5Glob` -> `FileGlob`
- `TestH5` -> `Subfile`
- `ExpectedH5` -> `ExpectedDataSubfile`

nilsvu commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,228 @@
+        config_lines = read_h5_checks_config_lines(self.input_filename)
+        parsed_yaml = yaml.safe_load(''.join(config_lines))
+        for check_block in parsed_yaml['H5Checks']:
+            h5_check_list += [
```
`h5_check_list.append(...)`

nilsvu commented on tools/CheckH5.py:
```diff
@@ -0,0 +1,228 @@
+        """
+        checks_performed = False
+        print(run_directory + "/" + self.h5_glob)
+        for filename in glob.glob(run_directory + "/" + self.h5_glob):
```
concatenate paths with `os.path.join(run_directory, self.h5_glob)`

nilsvu reviewed (changes requested):
This PR is doing _a lot_ of things that the PR title does not even hint at, like adding a dependency and the output file checking. Many of these things should be separate PRs. I don't feel qualified to "approve" the CCE commits since I sort of gave up checking all the fixes, math, actions and options they change, though I'm happy to review them in detail in separate PRs if you want me to. I _did_ review the H5 checking commit, so I'm happy for you to squash the fixups. Also I like the design for the output-file checking tool 👍

If this PR won't get factored into smaller PRs, please at least touch on the various things it does in the title and in the PR description. Also, I think it would be reasonable to add an "upgrade instructions" section that says pyyaml must be installed now.

moxcodes commented:
@nilsleiffischer Fair enough. I've made #2886.
I initially put all these changes together because I felt that the philosophy of 'don't merge code that isn't tested by automated checks' trumped 'keep PRs tidy', and the collection of changes here is only really verified by both having the h5 checking utility and having an input file to run with.

That said, I suppose the runs from this PR prove the thing works, so it seems fine to me if you prefer to only approve the h5 checking in the other PR.

moxcodes commented:
Rebased following merge of #2886.
Cheers!

nilsdeppe reviewed (dismissed):
Looks fine, but I have no idea if the fixes to the solutions are correct

moxcodes commented:
Thanks @nilsdeppe! The claim is that because the solutions are now run through an executable and the h5 checked that these should at least be 'more correct' :P

nilsdeppe commented:
@nilsleiffischer this is waiting on you :)

nilsvu commented on src/Evolution/Systems/Cce/Actions/InitializeFirstHypersurface.hpp:
```diff
@@ -35,9 +40,13 @@ namespace Actions {
   using const_global_cache_tags =
-      tmpl::list<Tags::LMax, Tags::NumberOfRadialPoints, Tags::InitializeJ>;
+      tmpl::list<Tags::LMax, Tags::NumberOfRadialPoints,
+                 std::conditional_t<
+                     tt::is_a_v<AnalyticWorldtubeBoundary, BoundaryComponent>,
+                     Tags::AnalyticInitializeJ, Tags::InitializeJ>>;
```
If I understand correctly, this decides how the `InitializeJ` tag ends up in the global cache, i.e. how it's read from the input file. I think this logic is better placed in the metavariables. Then this action just retrieves the `Tags::InitializeJ` from the box and doesn't care how it got in there.

moxcodes replied:
I don't think the metavariables is the right place either, though I see what you mean about preferring this action to be agnostic about the source.
I'll put it in the `const_global_cache_tags` of the worldtube boundary component -- that seems logically simplest.

nilsvu commented on src/Evolution/Systems/Cce/OptionTags.hpp:
```diff
@@ -506,6 +506,19 @@ struct InitializeJ : db::SimpleTag {
 
+// Tags that generates an `Cce::InitializeJ::InitializeJ` derived class from an
+// analytic solution.
+struct AnalyticInitializeJ : InitializeJ {
```
I don't think we have a simple tag that derives of another simple tag in the code so far. I'm not opposed to this pattern, but I remember @kidder and @nilsdeppe told me not to do this when we discussed how to add the elastic constitutive relation to the global cache. Here's the PR with that discussion: https://github.com/sxs-collaboration/spectre/pull/1915#discussion_r395602854. Have you changed your mind on this?

Note that these days the constitutive relation is always created as part of the analytic solution or data and a "reference tag" makes it available separately in the DataBox. Perhaps that's a pattern you can adopt as well.

nilsdeppe replied:
Still not supported and should be avoided. It might technically work, but is one of this "If it breaks we will probably just disable the code that uses that misfeature". I don't think there's a way to guard against it either, unfortunately. This is generally what "pure base tags" were meant to solve. Basically, simple tags can inherit from base tags only, compute tags must inherit from simple tags.

moxcodes replied:
In the fixup, I introduce a pure-base tag and inherit from it in the two separate cases -- I agree the inheriting from another simple tag is bad practice.

nilsdeppe replied:
:+1: sounds good :) Sorry, was just scrolling through the comments and realized there was a sort of ping to me here

nilsvu commented on src/Evolution/Systems/Cce/Actions/InitializeCharacteristicEvolutionScri.hpp:
```diff
@@ -39,10 +45,16 @@ namespace Actions {
+          tmpl::list<Tags::AnalyticBoundaryDataManager>, tmpl::list<>>>>;
+  using initialization_tags_to_keep = tmpl::conditional_t<
+      tt::is_a_v<AnalyticWorldtubeBoundary, BoundaryComponent>,
+      tmpl::list<Tags::AnalyticBoundaryDataManager>, tmpl::list<>>;
```
Why are these tags added to the DataBox, not the global cache?

moxcodes replied:
Because they are only used by a single component, so don't need to clutter the databox of every component.

nilsvu replied:
I see. If you want to move this logic to the parallel component as you did for `InitializeFirstHypersurface` then you could add this initialization tag in a separate action and have the parallel component decide whether to add it to the PDAL or not. Or just add it to the global cache along with `AnalyticInitializeJ`, just to keep things simple 🤷‍♂️

moxcodes replied:
It's also ill-suited to be in the global cache because it some analytic solutions need to modify internal state during their evolution ('analytic' here is used roughly -- the one in question actually numerically integrates a restricted set of variables).
I disagree that adding another action just to add the one tag is any simpler -- the analytic solution in this case is used during the scri observation action (because it needs to output the news from the analytic solution), so I think this initialization action is the right place for it.

nilsvu commented on src/Evolution/Systems/Cce/Actions/ScriObserveInterpolated.hpp:
```diff
@@ -202,6 +213,17 @@ struct ScriObserveInterpolated {
+        if constexpr (std::is_same_v<
+                          typename Metavariables::cce_boundary_component,
+                          AnalyticWorldtubeBoundary<Metavariables>>) {
+          db::get<Tags::AnalyticBoundaryDataManager>(box)
```
Maybe just check if the `Tags::AnalyticBoundaryDataManager` is in the `DbTags` instead of reaching up to the metavars?

nilsvu replied:
I think you can omit the first `if constexpr` and all the `has_cce_boundary_component` checking now

nilsvu commented on tests/InputFiles/Cce/AnalyticTestBouncingBlackHole.yaml:
```diff
@@ -0,0 +1,44 @@
+#   - Label: "check_news"
+#     Subfile: "/News.dat"
+#     FileGlob: "CharacteristicExtractVolume*.h5"
+#     ExpectedDataSubfile: "/News_expected.dat"
```
Do you think it would make sense here to output the _error_ to the analytic solution, instead of the analytic solution itself? This is what we do in the evolution and elliptic code.

moxcodes replied:
I dislike outputting only the error because it does not provide information about the scale of the expected solution, and therefore prevents usefully checking relative error. It seems like a reasonable practice if you know ahead of time that the solution is O(1), but that's not guaranteed for this case.

nilsvu replied:
You actually only use absolute tolerances in the input files currently, so should those perhaps switch to relative tolerances then? I'm actually not sure what's the right thing to do here. Your point on relative comparisons applies to evolution and elliptic errors just the same. I feel like it's more useful in practice to output errors, since that's what you typically want to plot and take norms of, plus it seems silly to output the same analytic data at every timestep / iteration. If you say we don't care that the CCE does this differently than the evolution and elliptic code then that would be fine with me. If we want to make it consistent though, then you could output errors here and keep your absolute tolerances in the input file unchanged. When writing tolerances in input files the author has to be aware of the scales and the expected accuracy of the solution anyway, don't they? What do you think?

moxcodes replied:
One important note here is that most of the CCE analytic solutions are not static solutions -- I see the point for elliptic solves, and for evolutions comparing to a static solution, but here it is valuable to know what the solution is actually doing when something happens with the errors.
I guess the elliptic solve and the evolution would handle this primarily by constructing the volume output as well when applicable, or adding some reduction data. 
However, given that the diagnostic of output data from the system is useful, then I don't see much of a philosophical difference in outputting the solution and errors to expected solution vs the solution and the expected solution directly.
When writing tolerances you need information about different things for absolute vs relative. For absolute you need the scale of the solution, which can vary depending on how complicated the analytic solution is -- we shouldn't anticipate that it's static in the evolution case. For relative tolerance, it depends on the convergence order and precision of the method, which I think should typically be more available.

nilsvu replied:
ok, I think this is ultimately your decision as the author of the code.

nilsvu commented on src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.hpp:
```diff
@@ -25,7 +26,74 @@ class ComplexDataVector;
+  static constexpr Options::String help = {
+      "Initialization process where J is set to analytic data associated with "
+      "the `LinearizedBondiSachs` analytic solution. This initial data "
+      "generator should not be used when not using the `LinearizedBondiSachs` "
```
remove double negative

nilsvu commented on src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.hpp:
```diff
@@ -25,7 +26,74 @@ class ComplexDataVector;
+///
+/// This initialization procedure should not be used except when the
+/// `Cce::Solutions::LinearizedBondiSachs` analytic solution is used.
+struct LinearizedBondiSachs : InitializeJ {
```
Could this commit be a follow-up PR? I find it very difficult to review a PR that mixes this sort of math-heavy code and the control-flow / "add an executable" code that is the rest of this PR. I think this PR could actually only add the test-executable with one input file that works, and then follow-up PRs can fix the individual analytic solutions. That would give the individual commits the attention they deserve :)

nilsvu commented on src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.hpp:
```diff
@@ -25,7 +26,74 @@ class ComplexDataVector;
+/// This initialization procedure should not be used except when the
+/// `Cce::Solutions::LinearizedBondiSachs` analytic solution is used.
+struct LinearizedBondiSachs : InitializeJ {
+  using options = tmpl::list<>;
```
Why no options? This seems to construct the class in an invalid state.

moxcodes commented:
Rebased following merge of #2903 and removed the more complicated linearized bondi sachs fix to be placed in a follow-up PR. 
Posted a fixup for the remaining comments related to the `specify InitializeJ` commit.
Cheers!

nilsdeppe commented:
fixup LGTM

nilsvu commented on tests/Unit/Evolution/Systems/Cce/Actions/Test_CharacteristicEvolutionBondiCalculations.cpp:
```diff
@@ -188,9 +189,9 @@ SPECTRE_TEST_CASE(
-      {start_time, l_max, number_of_radial_points,
-       std::make_unique<::TimeSteppers::RungeKutta3>(),
-       std::make_unique<InitializeJ::InverseCubic>()}};
+      {start_time, std::make_unique<InitializeJ::InverseCubic>(), l_max,
```
(optional, just a note in case you weren't aware of this) You can pass an explicit `tuples::TaggedTuple<TheTags...>{...}` to construct the runner, and it will reorder the tags for you. Then you don't have to care in which order the global cache tags happen to be.

nilsvu reviewed (commented):
The fixup LGTM, please squash it

nilsvu commented on src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp:
```diff
@@ -91,7 +87,7 @@ struct H5WorldtubeBoundary
   using base_type::initialize;
   using typename base_type::chare_type;
-  using typename base_type::const_global_cache_tag_list;
+  using const_global_cache_tags = tmpl::list<Tags::InitializeJ>;
```
These aliases should probably `tmpl::push_back` to the `base_type::const_global_cache_tag_list`

moxcodes replied:
the base currently does not have a `const_global_cache_tags` -- the one it had was incorrectly called `const_global_cache_tag_list` and didn't have anything the actions didn't have anyway, so isn't needed.

nilsvu replied:
probably better to delete it when you get the chance

moxcodes replied:
yep, the commit in question already does that. (lines 42-45)

moxcodes commented:
fixups squashed in, along with removing the has_boundary_component checking.
Cheers!

nilsvu commented on src/Evolution/Systems/Cce/Actions/ScriObserveInterpolated.hpp:
```diff
@@ -25,8 +25,13 @@
 #include "Utilities/MakeString.hpp"
 #include "Utilities/System/ParallelInfo.hpp"
 #include "Utilities/TMPL.hpp"
+#include "Utilities/TypeTraits/CreateHasTypeAlias.hpp"
```
Squash this line away :)

moxcodes replied:
oops... yep, done.
